src/xdg.c

```diff
@@ -114,17 +121,73 @@ handle_destroy(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, destroy);
 	assert(view->type == LAB_XDG_SHELL_VIEW);
+	struct xdg_toplevel_view *xdg_toplevel_view =
+		xdg_toplevel_view_from_view(view);
+
+	xdg_toplevel_view->xdg_surface->data = NULL;
+	xdg_toplevel_view->xdg_surface = NULL;
 
-	/* Reset XDG specific surface for good measure */
-	((struct xdg_toplevel_view *)view)->xdg_surface = NULL;
+	wl_list_remove(&view->map.link);
+	wl_list_remove(&view->unmap.link);
+	wl_list_remove(&view->request_move.link);
+	wl_list_remove(&view->request_resize.link);
+	wl_list_remove(&view->request_minimize.link);
+	wl_list_remove(&view->request_maximize.link);
+	wl_list_remove(&view->request_fullscreen.link);
+	wl_list_remove(&view->set_title.link);
```

Comment:
We could move the listener cleanup for the common `struct view` listeners to `view_destroy()` and only clean up the xdg specific ones here. This should likely be done in a separate PR though.

I like the `xdg_toplevel_view_from_view()` addition, inline casting like that always feels a bit meh.

Just a nit: I ended up using `struct xdg_toplevel_view *xdg_view = ..` so we don't have to spell out the variable name each time.